src/coreclr/jit/emit.h

```diff
@@ -2168,25 +2181,32 @@ class emitter
     insGroup* emitSavIG(bool emitAdd = false);
     void emitNxtIG(bool extend = false);
 
+    void emitRemoveLastInstruction();
+
     bool emitCurIGnonEmpty()
     {
         return (emitCurIG && emitCurIGfreeNext > emitCurIGfreeBase);
     }
 
     instrDesc* emitLastIns;
+    insGroup*  emitLastInsIG;
 
     // Check if a peephole optimization involving emitLastIns is safe.
     //
-    // We must have a lastInstr to consult.
+    // We must have a non-null emitLastIns to consult.
     // The emitForceNewIG check here prevents peepholes from crossing nogc boundaries.
     // The final check prevents looking across an IG boundary unless we're in an extension IG.
     bool emitCanPeepholeLastIns()
     {
-        return (emitLastIns != nullptr) &&                 // there is an emitLastInstr
-               !emitForceNewIG &&                          // and we're not about to start a new IG
-               ((emitCurIGinsCnt > 0) ||                   // and we're not at the start of a new IG
-                ((emitCurIG->igFlags & IGF_EXTEND) != 0)); //    or we are at the start of a new IG,
-                                                           //    and it's an extension IG
+        assert((emitLastIns == nullptr) == (emitLastInsIG == nullptr));
+
+        return (emitLastIns != nullptr) &&                   // there is an emitLastInstr
+               !emitForceNewIG &&                            // and we're not about to start a new IG
+               ((emitCurIGinsCnt > 0) ||                     // and we're not at the start of a new IG
+                ((emitCurIG->igFlags & IGF_EXTEND) != 0)) && //    or we are at the start of a new IG,
+                                                             //    and it's an extension IG
+               ((emitLastInsIG->igFlags & IGF_NOGCINTERRUPT) == (emitCurIG->igFlags & IGF_NOGCINTERRUPT));
```

Comment:
@BruceForstall : This change appears to be stopping some small instances of the existing IsRedundantLdStr optimisation due to the IGF_NOGCINTERRUPT flag not matching.

```
System.DirectoryServices.Protocols.LdapConnection:GetSaslDefaults():System.DirectoryServices.Protocols.SaslDefaultCredentials:this (MethodHash=65af3d8a)

G_M49781_IG02:        ; gcrefRegs=80000 {x19}, byrefRegs=100000 {x20}, byref
            stp     xzr, xzr, [fp, #0x18]
            stp     xzr, xzr, [fp, #0x28]
            str     xzr, [fp, #0x38]
            movz    x0, #0xD1FFAB1E
            movk    x0, #0xD1FFAB1E LSL #16
            movk    x0, #0xD1FFAB1E LSL #32
            str     x0, [fp, #0x18]
						;; size=28 bbWeight=1    PerfScore 5.50
G_M49781_IG03:        ; , nogc, extend
            ldr     x0, [fp, #0x18]. --------> this is not produced in HEAD, but is produced with this patch
            str     x0, [fp, #0x48]
            ldp     q16, q17, [fp, #0x20]
            stp     q16, q17, [fp, #0x50]
```

I'm not quite sure if this LDR is safe to remove or not.

Reply:
It might help to generate the disasm with `set COMPlus_JitDisasmWithGC=1`.

It might not be legal to remove the `ldr x0` if: (1) `x0` in the `str x0` is a gc ref, (2) the `x0` was marked as "gc dead" after the `str x0`, (3) a GC happens immediately after the `str`, such that `x0` no longer contains the correct value when the `str x0, [fp, #0x48]` is reached.

What I'm not positive about is: IG03 is a "nogc" region, so after the IG02 `str x0` we might be considered to be in that region and thus couldn't do the GC I mentioned in the step (3) above.

Of course, if `x0` is not a GC pointer, then it doesn't matter, and removing that `ldr x0` seems safe.

So arguably, the IGF_NOGCINTERRUPT check is too conservative. However, "nogc" regions are very rare, and we've had bugs (gc holes) with peephole optimizations crossing that boundary.

Reply:
Ok, let's leave it as it is in the patch. Better to be too conservative then unsafe. Plus, there are only 4 instances of IGF_NOGCINTERRUPT being hit across an entire SPMI run.